### PR TITLE
jgrss/coreg

### PR DESCRIPTION
**Pull request by jgrss**
## What is this PR changing?
This PR adds better CRS support for co-registration with AROSICS.

### CODE REVIEW

File: src/geowombat/core/api.py [hunk 1]
```diff
@@ -511,6 +526,19 @@ def __init__(
 
                     with xr.open_dataset(filename, **kwargs) as src:
                         self.data = src.to_array(dim='band')
+                    if band_names is not None:
```
Comment (jgrss) [inline src/geowombat/core/api.py:529]: This block allows the user to pass a desired band order after reading data from a NetCDF dataset (which might reorder the variables by name).

File: src/geowombat/core/sops.py [hunk 2]
```diff
@@ -1084,7 +1089,12 @@ def subset(
         return ds_sub
 
     @staticmethod
-    def coregister(target, reference, **kwargs):
+    def coregister(
+        target,
+        reference,
+        wkt_version: T.Optional[str] = None,
```
Comment (jgrss) [inline src/geowombat/core/sops.py:1095]: New argument to pass a WKT version when transforming a CRS.

File: src/geowombat/core/sops.py [hunk 3]
```diff
@@ -1116,66 +1127,103 @@ def coregister(target, reference, **kwargs):
...
-            cr.calculate_spatial_shifts()
-        except:
+        with tempfile.TemporaryDirectory() as tmp:
+            ref_path = Path(tmp) / '_reference.tif'
+            tar_path = Path(tmp) / '_target.tif'
+            with gw_.open(reference.filename, chunks={'band': -1, 'y': 256, 'x': 256}) as ref_src:
+                if wkt_version is not None:
+                    ref_src = ref_src.assign_attrs(
+                        {
+                            'crs': ref_src.gw.crs_to_pyproj.to_wkt(version=wkt_version)
```
Comment (jgrss) [inline src/geowombat/core/sops.py:1156]: Ensure temporary reference and target files have a WKT CRS.

File: src/geowombat/core/sops.py [hunk 4]
```diff
@@ -1116,66 +1127,103 @@ def coregister(target, reference, **kwargs):
...
+                        {
+                            'crs': tar_src.gw.crs_to_pyproj.to_wkt(version=wkt_version)
+                        }
+                    )
+                if 'nodata' in kwargs:
+                    tar_src = tar_src.assign_attrs({'nodatavals': (kwargs['nodata'][1],)}).fillna(kwargs['nodata'][1])
+                tar_src.gw.save(tar_path, overwrite=True)
+
+            cr = arosics.COREG(
+                GeoArray(GeoArray(str(ref_path)), projection=ref_src.crs),
```
Comment (jgrss) [inline src/geowombat/core/sops.py:1174]: The projection can only be set on a `geoarray.GeoArray` if another `GeoArray` is passed as the input. E.g.
```python
GeoArray(str(ref_path), projection=ref_src.crs)
```
would not pass the reference CRS but the following does.
```python
GeoArray(GeoArray(str(ref_path)), projection=ref_src.crs)
```

File: src/geowombat/core/sops.py [hunk 5]
```diff
@@ -1116,66 +1127,103 @@ def coregister(target, reference, **kwargs):
...
-                                               chunks=(target.gw.band_chunks,
-                                                       target.gw.row_chunks,
-                                                       target.gw.col_chunks)),
-                            dims=('band', 'y', 'x'),
-                            coords={'band': target.band.values.tolist(),
-                                    'y': ycoords,
-                                    'x': xcoords},
-                            attrs=target.attrs)
+        # Check if multi-band
+        if len(shift_info['arr_shifted'].shape) > 2:
```
Comment (jgrss) [inline src/geowombat/core/sops.py:1195]: This new logic allows single and multi-band co-reg.

File: tests/test_coreg.py [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+
+import geowombat as gw
+from geowombat.data import l8_224077_20200518_B2
+from geowombat.data import l8_224077_20200518_B4
+
+
+class TestCOREG(unittest.TestCase):
+    def test_coreg(self):
+        # with gw.config.update(ref_crs="epsg:8858"):
+        with gw.open(l8_224077_20200518_B2) as target, gw.open(l8_224077_20200518_B4) as reference:
```
Comment (jgrss) [inline tests/test_coreg.py:11]: Tests two UTM-projected images

File: tests/test_coreg.py [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+                resamp_alg_calc='cubic',
+                out_gsd=[30.0, 30.0],
+                q=True,
+                nodata=(0, 0),  # reference, data to be shifted
+                CPUs=1
+            )
+            self.assertTrue(reference.shape == data.shape)
+
+    def test_coreg_transform(self):
+        with gw.config.update(ref_crs="epsg:8858"):
```
Comment (jgrss) [inline tests/test_coreg.py:29]: Tests two images transformed to Equal Earth

File: tests/test_coreg.py [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+            with gw.open(l8_224077_20200518_B2, chunks=512) as target, \
+                gw.open(l8_224077_20200518_B4, chunks=512) as reference:
+                data = gw.coregister(
+                    target=target,
+                    reference=reference,
+                    wkt_version='WKT2_2019',
```
Comment (jgrss) [inline tests/test_coreg.py:35]: Pass a newer version of WKT so that Equal Earth is understood.

File: src/geowombat/core/api.py [hunk 9]
```diff
@@ -350,19 +350,33 @@ class open(object):
...
         >>>     print(src)
         >>>
+        >>> # Open a NetCDF image without access to transforms
+        >>> # NOTE: This will be faster than the above method
+        >>> # as it uses ``xarray.open_dataset`` and bypasses CRS checks.
+        >>> # NOTE: The chunks must be provided by the user.
+        >>> # NOTE: Providing band names will ensure the correct order when reading from a NetCDF dataset.
+        >>> with gw.open(
+        >>>     'image.nc',
+        >>>     chunks={'band': -1, 'y': 256, 'x': 256},
```
Comment (mmann1123) [inline src/geowombat/core/api.py:374]: Looks great. Is this covered in docs?

File: src/geowombat/core/sops.py (hunk 5 above)
Comment (mmann1123) [reply src/geowombat/core/sops.py:1203]: super cool

**Review by mmann1123 — APPROVED** [review]
Looks great. Just want to check that things get added to docs.  Maybe we should keep a check list of things that should be included?

File: src/geowombat/core/api.py (hunk 9 above)
Comment (jgrss) [reply src/geowombat/core/api.py:374]: No, but I like your idea of doing an inventory on methods and docs coverage.

**Pull request merged**